### PR TITLE
Changing the way decorators are stored

### LINKED ISSUE #395: Confusing behavior with class decorators
**Bug**

Base classes are getting decorators intended for subclasses.

**Code**

```typescript
function decorator(value: string) {
    return function (constructor: Function) {
        constructor.prototype.addDecorator(value);
    };
}

class Base {
    protected _decorators: any;

    protected addDecorator(value: string) {
        if (!this._decorators) {
            this._decorators = [];
        }

        this._decorators.push(value);
    }

    getDecorators() {
        return this._decorators;
    }
}

@decorator('Test1')
class Test1 extends Base {
}

@decorator('Test2')
class Test2 extends Test1 {
}

const test1 = new Test1();
const test2 = new Test2();

console.log("test1", test1.getDecorators());
console.log("test2", test2.getDecorators());
```

The logs reveal that `Test1` actually has two decorators now, instead of only one. The way this snippet works is similar to how `WidgetBase` adds decorators, which has the same problem.

**Expected behavior:**

`Test1` should have one decorator, `"Test1"`

**Actual behavior:**

`Test` has two decorators, `"Test2"`

### PR DESCRIPTION
**Pull request by rorticus**
**Type:** bug / feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

Fix for confusing class decorators

Resolves #395

### CODE REVIEW

File: src/WidgetBase.ts [hunk 1]
```diff
@@ -292,25 +292,68 @@ export class WidgetBase<P extends WidgetProperties> extends Evented implements W
 	 */
 	protected addDecorator(decoratorKey: string, value: any): void {
 		value = Array.isArray(value) ? value : [ value ];
-		if (!this._decorators) {
-			this._decorators = new Map<string, any[]>();
+
+		let decoratorList = decoratorMap.get(this.constructor);
```
Comment (agubler) [inline src/WidgetBase.ts]: nit: it's a Map not a list

File: src/WidgetBase.ts [hunk 2]
```diff
@@ -292,25 +292,68 @@ export class WidgetBase<P extends WidgetProperties> extends Evented implements W
...
+	 * @return An array of decorator values
+	 * @private
+	 */
+	private _buildDecoratorList(decoratorKey: string): any[] {
+		const allDecorators = [];
+
+		let constructor = this.constructor;
+
+		while (constructor) {
+			const decoratorList = decoratorMap.get(constructor);
```
Comment (agubler) [inline src/WidgetBase.ts]: nit: same as above, it's a map.

File: src/WidgetBase.ts (hunk 1 above)
Comment (rorticus) [inline src/WidgetBase.ts]: Updated!

File: src/WidgetBase.ts (hunk 2 above)
Comment (rorticus) [inline src/WidgetBase.ts]: Updated!

**Pull request merged**